crates/ecs/src/systems/mod.rs

```diff
@@ -408,7 +419,80 @@ impl<'a, Component> DerefMut for Write<'a, Component> {
     }
 }
 
+/// `Query` allows a system to access components from entities other than the currently iterated.
+///
+/// # Example
+/// ```
+/// # use ecs::filter::With;
+/// # use ecs::systems::{Query, Read, Write};
+/// # #[derive(Debug)] struct Mass(f32);
+/// # #[derive(Debug)] struct Velocity(f32);
+/// fn print_largest_momentum(query: Query<(Read<Mass>, Read<Velocity>)>) {
+///     let mut largest_momentum = 0.0;
+///     for (mass, velocity) in query {
+///         let momentum = mass.0 * velocity.0;
+///         if momentum > largest_momentum {
+///             largest_momentum = momentum;
+///         }
+///     }
+///     println!("The largest momentum of all entities is {largest_momentum} kg*m/s.");
+/// }
+/// ```
+#[derive(Debug)]
+pub struct Query<'world, P: SystemParameters> {
+    phantom: PhantomData<P>,
+    world: &'world World,
+}
+
+/// Iterator for [`Query`].
+#[derive(Debug)]
+pub struct QueryIterator<'components, P: SystemParameters> {
+    borrowed: P::BorrowedData<'components>,
+    iterate_over_entities: bool,
+    iterated_once: bool,
+}
```

Comment:
Do you think it's possible to implement [`std::ops::Index`](https://doc.rust-lang.org/std/ops/trait.Index.html) for the iterator (with type parameter `Idx = Entity`)? So that we'll be able to fetch component data for specific entities (i.e. the button component having the entity id of the door component to be able to find and open it), without needing to manually perform a linear search.

Reply:
I don't think it is possible to implement `std::ops::Index` because `index` returns `&Self::Output` and `fetch_parameter` will return an owned type. Instead, I could add a `get_entity` method on `Query` that does the same thing, but returns an owned type instead. This method can use the `entity_to_archetype_index` and `entity_to_component_index` hash-maps to fetch the parameters.

Reply:
Ah, that's too bad. Apparently there's [an RFC for this exact feature](https://github.com/rust-lang/rfcs/issues/997), but we'll have to make do without, then.

> Instead, I could add a `get_entity` method ...

Sounds good👍